dali/pipeline/operators/geometric/shapes.h

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATORS_GEOMETRIC_SHAPES_H_
+#define DALI_PIPELINE_OPERATORS_GEOMETRIC_SHAPES_H_
+
+#include <memory>
+#include <vector>
+#include "dali/kernels/tensor_shape.h"
+#include "dali/pipeline/operators/operator.h"
+#include "dali/core/static_switch.h"
+
+namespace dali {
+
+template <typename Backend>
+class Shapes : public Operator<Backend> {
+ public:
+  Shapes(const Shapes &) = delete;
+  explicit Shapes(const OpSpec &spec) : Operator<Backend>(spec) {
+    output_type_ = spec.GetArgument<DALIDataType>("type");
+    switch (output_type_) {
+    case DALI_INT32:
+    case DALI_UINT32:
+    case DALI_INT64:
+    case DALI_UINT64:
+    case DALI_FLOAT:
+    case DALI_FLOAT64:
+      break;
+    default:
+      {
+        auto &name = TypeTable::GetTypeInfo(output_type_).name();
+        DALI_FAIL("Operator Shapes can return the output as one of the following:\n"
+          "int32, uint32, int64, uint64, float or double;\n"
+          "requested: " + name);
+        break;
+      }
+    }
+  }
+  bool CanInferOutputs() const override { return true; }
+
+  bool SetupImpl(std::vector<OutputDesc> &output_desc, const workspace_t<Backend> &ws) override {
+    output_desc.resize(1);
+    output_desc[0].type = TypeTable::GetTypeInfo(output_type_);
+    decltype(auto) shape = GetInputShape(ws);
+    output_desc[0].shape = ShapeShape(shape);;
+    return true;
+  }
+
+  void RunImpl(workspace_t<Backend> &ws) override {
+    RunBackend(ws);
+  }
+
+  template <typename type>
+  void ConvertShape(TensorList<CPUBackend> &out, const kernels::TensorListShape<> &shape) {
+    assert(out.shape().num_samples() == shape.num_samples());
+    for (int i = 0; i < shape.num_samples(); i++) {
+      type *data = out.mutable_tensor<type>(i);
```

Comment:
In case of CPU you will do the allocation at this point implicitly.

Reply:
No?

Reply:
Irrelevant now.